### ISSUE TITLE
Rollups not calculating correctly over time

### ISSUE THREAD

**Issue by baobao917** [jamessimone/apex-rollup#624]
Hi @jamessimone - Not sure how exactly to demo this and perhaps it's been resolved with the latest release.  But I have noticed that over time the rollup calculations on records aren't 100% accurate.  This was noticed in version 1.6.13 and seen occurring in multiple orgs.

Essentially I instituted SUM rollups (both simple parent-child and also grandparent-child) and did a full recalc in the system and verified that the rollup calcs did indeed calculate correctly when comparing to the detail.  However, when I checked several months later the rollup calcs, it appeared that many of the target rollup fields did not have the correct result of the sum of the child detail.

I've updated to v.1.6.33 and have done a full recalc again to verify my starting point rollup calculations are correct and I guess I'll observe if it gets thrown out of whack again.  But I'm not sure why it would be or how to even begin giving you a starting point to troubleshoot.

**Comment by baobao917** [comment]
Quick follow-up - As a workaround, since I know the full recalc works, I've thought about doing a periodic weekly full recalc.  Is there a way to schedule the full recalc for a particular rollup? I didn't see it in the App

**Comment by jamessimone** [comment]
@baobao917 you can use a scheduled flow with the `Full Recalc CMDT-driven Invocable` invocable action - it takes a comma-separated list of the CMDT developer names and runs full recalcs for those rollups

**Comment by baobao917** [comment]
@jamessimone Thanks! I'll take a look at it.

**Comment by baobao917** [comment]
@jamessimone I had an issue with using the Full Recalc CMDT driven invocable.  I passed a comma separated list of 9 rollups (using the API names) and it's only doing full recalcs on 3 of them.  (It's not even the first 3 or last 3 in the list but just smack dab in the middle of the list).  Not sure if this is user error on my part or if it's a bug.  Also not sure if I should have spawned this off into a separate issue ticket.

Here is a video recording of the issue
https://www.loom.com/share/50814e63ab5d4820870076ae0fe70998

Pls let me know if you need anything else.

**Comment by baobao917** [comment]
> @jamessimone I had an issue with using the Full Recalc CMDT driven invocable. I passed a comma separated list of 9 rollups (using the API names) and it's only doing full recalcs on 3 of them. (It's not even the first 3 or last 3 in the list but just smack dab in the middle of the list). Not sure if this is user error on my part or if it's a bug. Also not sure if I should have spawned this off into a separate issue ticket.
> 
> Here is a video recording of the issue https://www.loom.com/share/50814e63ab5d4820870076ae0fe70998
> 
> Pls let me know if you need anything else.

@jamessimone I am going to open a separate issue for this issue with the Full Recalc CMDT driven invocable to separately track it from this original issue.

**Comment by jamessimone** [comment]
I won't be able to investigate this issue without more details. If you have Apex Rollup triggering on record insert/update/delete, all rollups will be kept in sync!

**Comment by baobao917** [comment]
> I won't be able to investigate this issue without more details. If you have Apex Rollup triggering on record insert/update/delete, all rollups will be kept in sync!

Hi @jamessimone - Yes, I'm not sure what details I can provide or how I can recreate.  The rollups when I first implement them calculate correctly.  I just have noticed in several client orgs where I am using this tool that when I come back to it several weeks/months later the results in the rollup target field aren't at all accurate.  I then either have to perform a full system recalculate (or click the individual record Recalculate Rollups button) to get the rollup target fields to reflect the proper results.  

I'd be happy to provide extra information if you can give me suggestions on what you need.  But since I can't pinpoint when/where the breakdown is occurring, I'm not sure how.

**Comment by jamessimone** [comment]
Thanks for the video. I'll do what I can to try to track this one down.

**Comment by baobao917** [comment]
Hi @jamessimone - I recorded another video below.  Last week I ran a full recalculation and validated all of the rollup fields reflected an accurate calculation.  It's been about a week and as you can see in the video, it is out of whack again.  Hopefully this helps.

https://www.loom.com/share/ac1afd8dee074919a4b38d94c2a57566?sid=db2c361b-3308-498d-af93-eebd41cc60dd

**Comment by jamessimone** [comment]
Thanks for the video! Question for you, after having watched it: when you're noticing these discrepancies, do the _amounts_ ever match up to recently updated children? I am going to play around with a few things to try to see what's going on.

**Comment by baobao917** [comment]
Hi James,

I'm not sure I understand the question. Are you asking if the target Rollup
field isn't calculating but rather is updating with the single value of a
recently updated child records?

**Comment by baobao917** [comment]
> Hi James,
> 
> I'm not sure I understand the question. Are you asking if the target Rollup
> field isn't calculating but rather is updating with the single value of a
> recently updated child records?
> […](#)

@jamessimone - Not sure if my response answered your question.  Pls let me know anything I can do to help.   If you need access to the org, please let me know.

**Comment by jamessimone** [comment]
@baobao917 is the `Is Full Record Set` checkbox toggled for these rollups?

**Comment by baobao917** [comment]
> [@baobao917](https://github.com/baobao917) is the `Is Full Record Set` checkbox toggled for these rollups?

@jamessimone Yes, for all of them (except one...which I probably just forgot).

![Image](https://github.com/user-attachments/assets/f6a628bd-a3cd-47c4-8366-60291a1dfc75)

![Image](https://github.com/user-attachments/assets/37cc6827-1ca9-4ae6-b5e7-4bec7e67f6c2)

**Comment by baobao917** [comment]
Hi @jamessimone - Following up on this to see if any progress has been made? It's still an issue.

**Comment by baobao917** [comment]
Hi @jamessimone - Have you had a chance to look at this?  It's still an issue.

**Comment by jamessimone** [comment]
Hey @baobao917, sorry for the delay in response. As far as I know, the only explanation for this behavior is as follows:

## First Potential Issue
* a scheduled full recalc operation is kicked off
* while that job is running, records associated with the rollup(s) in question are modified
* this can lead to inconsistencies

## Second Potential Issue
* a scheduled full recalc operation is kicked off
* some parent records fail to update due to any number of different reasons: record locks? validation rules? It could legitimately be anything
* this can lead to inconsistencies

I do not know of another mechanism that would cause drift beyond what I'm outlining here. Unfortunately, trying to address an issue like this can be quite difficult, because it really requires logging over time and being able to identify exactly when drift starts so that the logs can be gathered and examined. Failing that, I personally do not know of a way to proceed on this; if either of the hypotheses above is correct, there may not be a programmatic way for me to provide a fix, either.

**Comment by baobao917** [comment]
Hi @jamessimone - Thanks for the update.  In both potential issues, the common denominator is that a scheduled full recalc operation is kicked off.  I know that isn't happening b/c I'm the only person that would/could do that.  And I'm unable to kick off a scheduled recalc b/c per issue #629 , I've never been able to get that to work.

Yes I understand the difficulty. I couldn't figure out a way to recreate or pinpoint the cause myself.  I know the issue isn't validation rules b/c when I do a manual full recalc, everything calculates correctly.  It's a shame b/c that just means we're not really able to rely on Apex Rollup to give us accurate results.
